picky/src/jose/jwe.rs

```diff
@@ -455,58 +622,78 @@ enum EncoderMode<'a> {
-    let protected_header_base64 = general_purpose::URL_SAFE_NO_PAD.encode(serde_json::to_vec(&header)?);
+fn encode_impl(mut jwe: Jwe, mode: EncoderMode) -> Result<String, JweError> {
+    use picky_asn1_x509::PublicKey as RfcPublicKey;
 
     let (encrypted_key_base64, jwe_cek) = match mode {
         EncoderMode::Direct(symmetric_key) => {
-            if symmetric_key.len() != header.enc.key_size() {
+            if symmetric_key.len() != jwe.header.enc.key_size() {
                 return Err(JweError::InvalidSize {
                     ty: "symmetric key",
-                    expected: header.enc.key_size(),
+                    expected: jwe.header.enc.key_size(),
                     got: symmetric_key.len(),
                 });
             }
 
             // Override `alg` header with "dir"
-            header.alg = JweAlg::Direct;
+            jwe.header.alg = JweAlg::Direct;
 
-            (String::new(), Cow::Borrowed(symmetric_key))
-        }
-        EncoderMode::Asymmetric(public_key) => {
-            // Currently, only rsa is supported
-            let rsa_public_key = RsaPublicKey::try_from(public_key)?;
-
-            let mut rng = rand::rngs::OsRng;
-
-            let mut symmetric_key = vec![0u8; header.enc.key_size()];
-            rng.fill_bytes(&mut symmetric_key);
-
-            let padding = match header.alg {
-                JweAlg::RsaPkcs1v15 => PaddingScheme::new_pkcs1v15_encrypt(),
-                JweAlg::RsaOaep => PaddingScheme::new_oaep::<sha1::Sha1>(),
-                JweAlg::RsaOaep256 => PaddingScheme::new_oaep::<sha2::Sha256>(),
-                unsupported => {
-                    return Err(JweError::UnsupportedAlgorithm {
-                        algorithm: format!("{:?}", unsupported),
-                    })
-                }
-            };
-
-            let encrypted_key = rsa_public_key.encrypt(&mut rng, padding, &symmetric_key)?;
-
-            (
-                general_purpose::URL_SAFE_NO_PAD.encode(encrypted_key),
-                Cow::Owned(symmetric_key),
-            )
+            (String::new(), Zeroizing::new(symmetric_key.to_vec()))
         }
+        EncoderMode::Asymmetric(public_key) => match &public_key.as_inner().subject_public_key {
+            RfcPublicKey::Rsa(_) => {
+                let rsa_public_key = RsaPublicKey::try_from(public_key)?;
+
+                let padding = match jwe.header.alg {
+                    JweAlg::RsaPkcs1v15 => PaddingScheme::new_pkcs1v15_encrypt(),
+                    JweAlg::RsaOaep => PaddingScheme::new_oaep::<sha1::Sha1>(),
+                    JweAlg::RsaOaep256 => PaddingScheme::new_oaep::<sha2::Sha256>(),
+                    unsupported => {
+                        return Err(JweError::UnsupportedAlgorithm {
+                            algorithm: format!("{:?}", unsupported),
+                        })
+                    }
+                };
+
+                let cek = generate_cek(jwe.header.enc);
```

Comment:
Moved symmetric key generation to one place, and now I maximized use of `zeroize::Zeroizing` to clear keys on drop